File: services/typescript/express/src/routes/handlers/delete.ts
import type { Request, RequestHandler, Response } from "express";
import { parseId, type State } from "app";
import { getLogger } from "telemetry";
import { AppErrors, sendError } from "../errors.js";

export function deleteHandler(state: State): RequestHandler {
  return (req: Request, res: Response) => {
    const logger = getLogger();

    const id = parseId(req.params.id);
    if (!id.success) {
      logger.warn({ method: req.method, path: req.path }, "Invalid path");
      return sendError(res, AppErrors.InvalidPath);
    }

    const task = state.tasks.get(id.data);

    if (task === undefined) {
      logger.warn(
        { id: id.data, method: "DELETE", path: req.path },
        "Task not found"
      );
      return sendError(res, AppErrors.TaskNotFound);
    }

    state.tasks.delete(id.data);

    logger.info(
      {
        id: id.data,
        secret: task.secret.length,
        operation: task.operation.toString().toLowerCase(),
        method: "DELETE"
      },
      "Removed task"
    );

    return res.status(204).send();
  };
}

File: services/python/quart/src/routes/handlers/delete.py
import structlog
from uuid import UUID
from quart import current_app, request
from pydantic import ValidationError

from app.state import AppState
from app.params import parse_id
from app.task import Task
from routes.errors import send_error, AppErrors


async def delete_handler(id: str):
    logger = structlog.get_logger()

    try:
        task_id: UUID = parse_id(id)
    except ValidationError:
        logger.warn("Invalid path", method=request.method, path=request.path)
        return send_error(AppErrors.InvalidPath)

    state: AppState = current_app.extensions["state"]
    task: Task | None = state.tasks.pop(task_id, None)

    if task is None:
        logger.warn(
            "Task not found", id=str(task_id), method=request.method, path=request.path
        )
        return send_error(AppErrors.TaskNotFound)

    logger.info(
        "Removed task",
        id=str(task_id),
        secret=len(task.secret),
        operation=task.operation.lower(),
        method=request.method,
    )

    return "", 204
